$NetBSD: patch-CMakeLists.txt,v 1.2 2019/06/19 22:39:55 bergerspencer Exp $

Solaris ld: Provided assembler is not PIC, so we pass "-mimpure-text"
            to the compiler so it doesn't send "-ztext" to ld.

Solaris ld: Passing "-M /usr/lib/ld/map.noexstk" to ld will generate a
            PT_SUNWSTACK ELF header that sets the stack permissions to
            rw. May not matter much since this is the default for 64
            bit executables, but include the flag just in case.

Solaris ld: Unlike the GNU ld, the Solaris ld doesn't ignore when
            -Bsymbolic is passed to something other than a shared lib
            and throws an error. Regression in
            https://bitbucket.org/multicoreware/x265/commits/83bc6fac1fb54e9d5241c5c10d8578811a355273
--- CMakeLists.txt.orig	2019-01-23 09:47:18.000000000 +0000
+++ CMakeLists.txt
@@ -620,7 +620,12 @@ if(ENABLE_SHARED)
         elseif(CYGWIN)
             # Cygwin is not officially supported or tested. MinGW with msys is recommended.
         else()
-            list(APPEND LINKER_OPTIONS "-Wl,-Bsymbolic,-znoexecstack")
+            if (${CMAKE_SYSTEM_NAME} MATCHES "SunOS")
+                list(APPEND LINKER_OPTIONS "-mimpure-text -Wl,-M,/usr/lib/ld/map.noexstk")
+            else()
+                list(APPEND LINKER_OPTIONS "-Wl,-znoexecstack")
+            endif()
+            list(APPEND SHARED_LINKER_OPTIONS "-Wl,-Bsymbolic")
         endif()
     endif()
     set_target_properties(x265-shared PROPERTIES SOVERSION ${X265_BUILD})
@@ -641,8 +646,8 @@ if(ENABLE_SHARED)
     endif()
     if(LINKER_OPTIONS)
         # set_target_properties can't do list expansion
-        string(REPLACE ";" " " LINKER_OPTION_STR "${LINKER_OPTIONS}")
-        set_target_properties(x265-shared PROPERTIES LINK_FLAGS "${LINKER_OPTION_STR}")
+        string(REPLACE ";" " " SHARED_LINKER_OPTION_STR "${LINKER_OPTIONS} ${SHARED_LINKER_OPTIONS}")
+        set_target_properties(x265-shared PROPERTIES LINK_FLAGS "${SHARED_LINKER_OPTION_STR}")
     endif()
 endif()
 
